test/gtest/ternary_tensor_ops.cpp

```diff
@@ -0,0 +1,452 @@
+    }
+
+private:
+    tensor<T> tensorA;
+    tensor<T> tensorB;
+    tensor<T> tensorC;
+};
+
+struct GPU_TensorOps_FP32 : public TensorOpsCommon<float>
+{
+};
+
+struct GPU_TensorOps_FP16 : public TensorOpsCommon<half_float::half>
+{
+};
+
+struct GPU_TensorOps_FP64 : public TensorOpsCommon<double>
+{
+};
+
+bool checkTensorsCompatibility(const std::vector<int>& tensorALens,
+                               const std::vector<int>& tensorBLens)
+{
+    if(tensorALens.size() != tensorBLens.size())
+    {
+        return false;
+    }
+
+    for(size_t idx = 0; idx < tensorBLens.size(); ++idx)
+    {
+        if((tensorBLens[idx] != 1) && (tensorALens[idx] != tensorBLens[idx]))
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+template <typename T>
+void AddTestCases(std::vector<TestCase>& testCases,
+                  const std::vector<int> tensorALens,
+                  const std::vector<int> tensorBLens)
+{
+    const auto& stride_a = stridesArr[0];
+    const auto& stride_b = stridesArr[0];
+    const auto& stride_c = stridesArr[0];
+
+    for(bool packed : packedArr)
+        for(const auto& offsets : offsetsArr)
+        {
+            std::vector<int64_t> final_offsets{0, 0, 0};
+            if(!packed)
+            {
+                if(std::any_of(offsets.begin(), offsets.end(), [](int64_t o) { return o < 0; }))
+                    continue;
+
+                final_offsets = offsets;
+            }
+
+            auto checkStride = [p = packed](const std::vector<int>& lens,
+                                            const std::vector<int>& strides) {
+                if(p)
+                    return true;
+
+                if(lens.size() > strides.size())
+                    return false;
+
+                // only sparsed case allowed, since all the kernels do not support the last
+                // dimension strides
+                if(strides.back() == 1)
+                {
+                    auto packedStrides =
+                        miopen::TensorDescriptor(miopen_type<T>{}, lens).GetStrides();
+                    return std::equal(packedStrides.rbegin(),
+                                      packedStrides.rend(),
+                                      strides.rbegin(),
+                                      [](int ps, int s) { return s >= ps; });
+                }
+
+                // currently tensor operations do not support non-one stride in the last dimention.
+                return false;
+            };
+
+            if(!checkStride(tensorALens, stride_a))
+                continue;
+            if(!checkStride(tensorBLens, stride_b))
+                continue;
+            if(!checkStride(tensorALens, stride_c))
+                continue;
+
+            for(const auto& alphabeta : alphabetaArr)
+                for(const auto& operation : operationArr)
+                {
+                    TestCase& testCase = testCases.emplace_back();
```

Comment:
It's better to properly `emplace` everything, passing all the parameters into the `emplace` method.